rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,345 @@
+
+#### Compute
+
+The following operators have a fixed input and output type of `Arrow`. We can
+group them into *filter* operators that do not change the schema, and *reshape*
+operators where execution yields a new schema.
+
+##### Filter
+
+Filter operators have input and output type `Arrow`.
+
+| Operator | Non-Blocking | Description
+| -------- |:------------:| ------------------------------------------
+| `head`   | ✅           | Retrieve first N records
+| `tail`   | ❌           | Retrieve last N records
+| `where`  | ✅           | Filter with expression
+| `sort`   | ❌           | Sort according to field
+| `uniq`   | ✅           | Produce deduplicated output
+
+##### Reshape
+
+Reshape operators have input and output type `Arrow`.
+
+| Operator    | Non-Blocking | Description
+| ----------- |:------------:| ------------------------------------------
+| `put`       | ✅           | Select a set of columns (projection)
+| `with`      | ✅           | Adds new fields with derived values
+| `drop`      | ✅           | Remove a set of columns (projection)
+| `rename`    | ✅           | Renames schema meta data (type & fields)
+| `summarize` | ❌/✅[^1]    | Aggregate group and compute a summary function
+
+[^1]: Depending on the aggregate function.
+
+##### Matcher Plugin
+
+| Operator | Input Type | Output Type | Description
+| -------- | ---------- | ----------- | ------------------------------------------
+| `build`  | `Arrow`    | `Matcher`   | Creates a matcher from a data stream
+| `match`  | `Arrow`    | `Arrow`     | Matches input based on its state
+
+#### Commands
+
+The following table illustrates how the current CLI would change after this
+proposal:
+
+| Old Syntax                               | New Syntax
+| ---------------------------------------- | -------------------------------------------
+| `vast infer`                             | `vast exec 'read [FORMAT] | infer"`
+| `vast import FORMAT EXPR`                | `vast import FORMAT 'EXPR | op1 .. | op2 ..'`
+| `vast export FORMAT EXPR`                | `vast export FORMAT 'EXPR | op1 .. | op2 ..'`
+| `vast count [args] EXPR`                 | `vast export FORMAT 'EXPR | count [args]'`
+| `vast pivot --format=FORMAT [args] EXPR` | `vast export FORMAT 'EXPR | pivot [args]'`
+| `vast explore [args] EXPR`               | `vast export 'EXPR | explore [-A] [...]'`
+
+In essence, we keep `import` and `export` and incrementally enhance the
+expressiveness, by making it possible to append a pipeline after an expression.
+That is, valid input has the form `EXPR | OPERATOR | OPERATOR | ...`. Note that
+this is *not* a valid pipeline, because `EXPR` is not an operator.
+
+Additionally, we add a new command `exec` that accepts a "pure" pipeline, i.e.,
+just a composition of operators. Execution takes place locally, without any
+implicit remote communication. For the remainder of the discussion, we focus
+on `exec` only.
```

Comment:
This should explain that pipelines usable by `exec` must have the input and output type Void.

Reply:
Explained in 6a9b75fe0.